fix(oxc_transform): overlap replacement

IWANABETHATGUY opened the pull request:
closed  https://github.com/oxc-project/oxc/issues/7594

### Reference
https://github.com/oxc-project/oxc/pull/7343/files

IWANABETHATGUY commented:
* **#7621** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7621?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/7621?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

IWANABETHATGUY commented:
Since `TraverseCtxt` has no functionality to skip visit sub tree(Fix me if I am wrong, I don't find the usage in codebase).
Need to lock the sub tree if one ast node is replaced by some value. 
TODO: 
if overlap appears in `AssignmentExpr` the output should still be wrong, since we can't get address of `AssignmentExpression` for now. 

@overlookmotel any chance to implement `ToAddress` for `AssignmentExpression` ?

overlookmotel commented:
> Since `TraverseCtxt` has no functionality to skip visit sub tree(Fix me if I am wrong, I don't find the usage in codebase). 

Correct. `Traverse` has no facility to skip visiting sub tree at present. We could look at adding that, but it might cause quite a performance regression, as it'd introduce an extra check and branch after every single AST node.

> TODO: if overlap appears in `AssignmentExpr` the output should still be wrong, since we can't get address of `AssignmentExpression` for now.
> 
> @overlookmotel any chance to implement `ToAddress` for `AssignmentExpression` ?

I'm not familiar with the define plugin, so not sure what you mean by "overlap". Can you give an example showing the problem with `AssignmentExpression`?

IWANABETHATGUY commented:
> > Since `TraverseCtxt` has no functionality to skip visit sub tree(Fix me if I am wrong, I don't find the usage in codebase).
> 
> Correct. `Traverse` has no facility to skip visiting sub tree at present. We could look at adding that, but it might cause quite a performance regression, as it'd introduce an extra check and branch after every single AST node.
> 
> The way you are handling it here seems like a good option in meantime, especially as the subtree is likely small, so not a great cost to visiting it.
> 
> > TODO: if overlap appears in `AssignmentExpr` the output should still be wrong, since we can't get address of `AssignmentExpression` for now.
> > @overlookmotel any chance to implement `ToAddress` for `AssignmentExpression` ?
> 
> I'm not familiar with the define plugin, so not sure what you mean by "overlap". Can you give an example showing the problem with `AssignmentExpression`?

Here is one, https://esbuild.github.io/try/#dAAwLjI0LjAAewogIGRlZmluZTogewogICAgJ2ltcG9ydC5tZXRhLmVudi5GT08nOiAnaW1wb3J0Lm1ldGEuZW52LkZPTycsCiAgICAnaW1wb3J0Lm1ldGEuZW52JzogJ19fZm9vX18nLAogIH0sCn0ACmNvbnNvbGUubG9nKAoJW2ltcG9ydC5tZXRhLmVudi5GT08gPSAwXQopCg, you could see 
both 
`import.meta.env.FOO` and `import.meta.env` matched part of  `import.meta.env.FOO`
for esbuild it will replace `import.meta.env.FOO` to `import.meta.env.FOO` and skip visit sub tree. 

but for us, 
we first replace `import.meta.env.FOO` to `import.meta.env.FOO`, then visit the child `import.meta.env`, and replace it to 
`__foo__`, so the final output should be `__foo__.FOO`

overlookmotel commented on crates/oxc_transformer/src/plugins/replace_global_defines.rs:
```diff
@@ -220,26 +220,55 @@ pub struct ReplaceGlobalDefinesReturn {
+            return;
+        }
+        if self.replace_define_with_assignment_expr(node, ctx) {
+            // `AssignmentExpression` is always stored in a `Box`, since it came from Traverse Visitor
```
It is true that `AssignmentExpression` is always stored in a `Box`, but the reason for that isn't that it came from the `Traverse` visitor. The reason is that the only place `AssignmentExpression` appears in AST types is in:

```rs
enum Expression<'a> {
  // ...
  AssignmentExpression(Box<'a, AssignmentExpression<'a>>),
  // ...
}
```

This doesn't generalize to all types that come from the `Traverse` visitor. e.g. `IdentifierName` isn't always stored in a `Box` (e.g. in `StaticMemberExpression`).

Not trying to lecture you! Just explaining, as I know you're interested in using `GetAddress` in Rolldown, and it's useful to know its limitations/gotchas.

IWANABETHATGUY replied:
thanks

overlookmotel commented on crates/oxc_transformer/tests/integrations/plugins/replace_global_defines.rs:
```diff
@@ -68,7 +67,22 @@ fn dot_with_overlap() {
     ])
     .unwrap();
     test("import.meta.env", "__foo__", config.clone());
-    test("import.meta.env.NODE_ENV", "import.meta.env.NODE_ENV", config.clone());
```
Shouldn't we keep this test, but change it's output?

```rs
test("import.meta.env.NODE_ENV", "__foo__.NODE_ENV", config.clone());
```

IWANABETHATGUY replied:
Done

overlookmotel commented on crates/oxc_transformer/tests/integrations/plugins/replace_global_defines.rs:
```diff
@@ -68,7 +67,22 @@ fn dot_with_overlap() {
-    test("import.meta.env.NODE_ENV", "import.meta.env.NODE_ENV", config.clone());
+    test("import.meta.env.FOO", "import.meta.env.FOO", config.clone());
+
+    test(
+        r"
+console.log(
+	[import.meta.env.FOO = 0]
+)
+        ",
+        "
+
+console.log(
+	[import.meta.env.FOO = 0]
+)
+",
+        config.clone(),
+    );
```
This test seems overly complicated, and doesn't cover all cases.

```suggestion
    test("import.meta.env = 0", "__foo__ = 0");
    test("import.meta.env.NODE_ENV = 0", "__foo__.NODE_ENV = 0");
    test("import.meta.env.FOO = 0", "import.meta.env.FOO = 0");
```

IWANABETHATGUY replied:
Yes

overlookmotel commented on crates/oxc_transformer/src/plugins/replace_global_defines.rs:
```diff
@@ -220,26 +220,55 @@ pub struct ReplaceGlobalDefinesReturn {
+        if self.ast_node_lock.is_some() {
+            return;
+        }
+        let mut flag = self.replace_identifier_defines(expr, ctx);
+        flag |= self.replace_dot_defines(expr, ctx);
+        if flag {
```
```suggestion
        let is_replaced = self.replace_identifier_defines(expr, ctx)
            || self.replace_dot_defines(expr, ctx);
        if is_replaced {
```

overlookmotel reviewed (changes requested):
Apart from the comments I've made, looks good.

But could you please add some comments about what `ast_node_lock` is and how it works?

overlookmotel commented:
### Merge activity

* **Dec 6, 10:26 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

IWANABETHATGUY committed:
fix(oxc_transform): overlap replacement (#7621)

closed  https://github.com/oxc-project/oxc/issues/7594

### Reference
https://github.com/oxc-project/oxc/pull/7343/files
